File: Python/Problems/Problems/problem1432.py
from .problem import Problem

class Problem1432(Problem):
    def test(self):
        num = 123456

        expected = 820000

        result = self.maxDiff(num)

        print(result)

        return expected == result

    def maxDiff(self, num):
        numsString = str(num)
        swapToMin = 0
        swapToMax = 0

        while swapToMin < len(numsString):
            if numsString[swapToMin] != '0' and numsString[swapToMin] != '1':
                break
            swapToMin += 1

        while swapToMax < len(numsString):
            if numsString[swapToMax] != '9':
                break
            swapToMax += 1

        max = numsString
        min = numsString
        if swapToMin < len(numsString):
            swapToMinC = numsString[swapToMin]
            min = min.replace(swapToMinC, ('1' if swapToMinC == numsString[0] else '0'))

        if swapToMax < len(numsString):
            swapToMaxC = numsString[swapToMax]
            max = max.replace(swapToMaxC, '9')

        return int(max) - int(min)

File: C++/Problems/Problems/problem_1432.cpp
#include "problem_1432.h"

#include <algorithm>
#include <iostream>
#include <ostream>

bool problem_1432::test() {
    auto num = 555;

    auto expected = 888;

    auto result = maxDiff(num);

    std::cout << result << std::endl;

    return expected == result;
}

int problem_1432::maxDiff(int num) {
    auto numsString = std::to_string(num);
    int swapToMin = 0;
    int swapToMax = 0;

    for (; swapToMin < numsString.size(); swapToMin++) {
        if (numsString[swapToMin] != '0' && numsString[swapToMin] != '1')
            break;
    }

    for (; swapToMax < numsString.size(); swapToMax++) {
        if (numsString[swapToMax] != '9')
            break;
    }
    std::string max = numsString;
    std::string min = numsString;
    auto swapToMinC = numsString[swapToMin];
    auto swapToMaxC = numsString[swapToMax];
    if (swapToMin < numsString.size())
        std::replace(min.begin(), min.end(), swapToMinC, swapToMinC == numsString[0] ? '1' : '0');

    if (swapToMax < numsString.size())
        std::replace(max.begin(), max.end(), swapToMaxC, '9');


    return std::stoi(max) - std::stoi(min);
}
